connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamingQueryCache.scala

```diff
@@ -0,0 +1,209 @@
+ *   - Once a query is stopped, the reference and mappings are maintained for 1 hour and will be
+ *     accessible from the client. This allows time for client to fetch status. If the client
+ *     continues to access the query, it stays in the cache until 1 hour of inactivity.
+ *
+ * Note that these semantics are evolving and might change before being finalized in Connect.
+ */
+private[connect] class SparkConnectStreamingQueryCache(
+    val sessionKeepAliveFn: (String, String) => Unit, // (userId, sessionId) => Unit.
+    val clock: Clock = new SystemClock(),
+    private val stoppedQueryInactivityTimeout: Duration = 1.hour, // Configurable for testing.
+    private val sessionPollingPeriod: Duration = 1.minute // Configurable for testing.
+) extends Logging {
+
+  import SparkConnectStreamingQueryCache._
+
+  def registerNewStreamingQuery(sessionHolder: SessionHolder, query: StreamingQuery): Unit = {
+    queryCacheLock.synchronized {
+      val value = QueryCacheValue(
+        userId = sessionHolder.userId,
+        sessionId = sessionHolder.sessionId,
+        session = sessionHolder.session,
+        query = query,
+        expiresAtMs = None)
+
+      queryCache.put(QueryCacheKey(query.id.toString, query.runId.toString), value) match {
+        case Some(existing) => // Query is being replace. Not really expected.
+          log.warn(
+            s"Replacing existing query in the cache (unexpected). Query Id: ${query.id}." +
+              s"Existing value $existing, new value $value.")
+        case None =>
+          log.info(s"Adding new query to the cache. Query Id ${query.id}, value $value.")
+      }
+
+      schedulePeriodicChecks() // Starts the scheduler thread if it hasn't started.
+    }
+  }
+
+  /**
+   * Returns [[StreamingQuery]] if it is cached and session matches the cached query. It ensures
+   * the the session associated with it matches the session passed into the call. If the query is
+   * inactive (i.e. it has a cache expiry time set), this access extends its expiry time. So
+   * if a client keeps accessing a query, it stays in the cache.
+   */
+  def getCachedQuery(queryId: String, runId: String, session: SparkSession):
+  Option[StreamingQuery] = {
+    val key = QueryCacheKey(queryId, runId)
+    queryCacheLock.synchronized {
+      queryCache.get(key).flatMap { v =>
+        if (v.session == session) {
+          v.expiresAtMs.foreach { _ =>
+            // Extend the expiry time as the client is accessing it.
+            val expiresAtMs = clock.getTimeMillis() + stoppedQueryInactivityTimeout.toMillis
+            queryCache.put(key, v.copy(expiresAtMs = Some(expiresAtMs)))
+          }
+          Some(v.query)
+        }
+        else None // This should be rare, may be client is trying access from a different session.
+      }
+    }
+  }
+
+  // Visible for testing
+  private[service] def cacheQueryValue(queryId: String, runId: String): Option[QueryCacheValue] =
```

Comment:
nit: this function name should be `getCacheQueryValue` ?

Reply:
Rename it to `getCachedValue()`. This is only accessed in tests.